lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs

```diff
@@ -203,3 +237,123 @@ spec = do
             let msg = "Address has unexpected length 1: Address {unAddress = \"0\"}"
             evaluate (runPut $ putAddress $ Address "0")
                 `shouldThrow` userException msg
+
+        it "decode (encode address) === address" $ property $ \addr ->
+            let
+                bytes = (runPut $ putAddress addr)
+                _hex = convertToBase @ByteString @ByteString Base16 (BL.toStrict bytes)
+            in
+                counterexample ("bytes: " <> show bytes) $ do
+                let addr' = try' $ runGet getAddress bytes
+                addr' === (Right addr)
+
+        it "decode (encode tx) === (validate tx)" $ property $ \(SignedTx signedTx) ->
+            let
+                bytes = try' (runPut $ putSignedTx signedTx)
+                hex = convertToBase @ByteString @ByteString Base16 . BL.toStrict <$> bytes
+                numberOfIns = length . inputs . fst $ signedTx
+                numberOfOuts = length . outputs . fst $ signedTx
+            in
+                counterexample ("bytes: " <> show hex) $
+                tabulate "number of inputs" [show numberOfIns] $
+                counterexample (show (numberOfIns, numberOfOuts)) $ do
+                let tx' = try' . unMessage . runGet getMessage =<< bytes
+                labelTx signedTx $ tx' === validate signedTx
+  where
+    unMessage :: Message -> (Tx, [TxWitness])
+    unMessage m = case m of
+        Transaction stx -> stx
+        _ -> error "expected a Transaction message"
+
+    try' :: a -> Either String a
+    try' = unsafePerformIO . wrap
+
+    wrap :: a -> IO (Either String a)
+    wrap = fmap (either (Left . show) Right) . (try @SomeException) . evaluate
```

Comment:
The two functions above are dodgy :s ... What's wrong with using `runGet` ?

Reply:
> What's wrong with using runGet

Not sure what you're referring to here, but `runGet` throws exceptions. And it seemed easier to not have to use the different api of `Test.QuickCheck.Monadic` when I was starting this.

There *is* actually an `Either`-returning version of `runGet`. But none for `runPut`.